### PR TITLE
Update way handling of DataTemplateSelector with null values

### LINKED ISSUE #2170: ItemsRepeater crashes when using custom ItemTemplateSelector that does not implement SelectTemplateCore(item) override
**Describe the bug**
I want to create a combination of treeview/listview that would behave like this: have a few roots (such as: Effects/Transitions/Text Effects). Each will expand to a few effect packs. An effect pack will show a (wrap panel) list of effects.

I wanted to implemement this as an ItemsRepeater, but no matter what I do, it crashes at startup (the app can't even start).

**Steps to reproduce the bug**
It could be that I'm doing something wrong, but I don't know what that would be. I've trimmed off pretty much everything, just to make it easy to debug.

	<UserControl
		x:Class="gridsplittertest.browse_effects_ctrl"
		xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
		xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
		xmlns:local="using:gridsplittertest"
		xmlns:muxc="using:Microsoft.UI.Xaml.Controls"
		xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
		xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
		mc:Ignorable="d"
		d:DesignHeight="300"
		d:DesignWidth="400">
		<UserControl.DataContext>
			<local:browse_effects_view_model/>
		</UserControl.DataContext>
		<UserControl.Resources>

			<DataTemplate x:Key="et" x:DataType="local:effect_type">
				<Grid Height="10" Width="200" MaxHeight="10" MaxWidth="10">
				</Grid>
			</DataTemplate>

			<DataTemplate x:Key="ep" x:DataType="local:effect_pack">
				<Grid Height="10" Width="200" MaxHeight="10" MaxWidth="10">
				</Grid>

			</DataTemplate>

			<DataTemplate x:Key="ei" x:DataType="local:effect_info">
				<Grid></Grid>
			</DataTemplate>

			<local:effect_data_template_selector x:Key="edts" effect_type_template="{StaticResource et}" effect_pack_template="{StaticResource ep}" effect_item_template="{StaticResource ei}"
			/>

		</UserControl.Resources>
		<Grid>
			<muxc:ItemsRepeater ItemsSource="{Binding ui_effect_types}" ItemTemplate="{StaticResource edts}">
				<muxc:ItemsRepeater.Layout>
					<muxc:StackLayout Orientation="Vertical"/>
				</muxc:ItemsRepeater.Layout>
			</muxc:ItemsRepeater>
		</Grid>
	</UserControl>

To my knowledge, the above should work. Am I doing something wrong?

I've attached the whole sample project I'm using.
[gridsplittertest.zip](https://github.com/microsoft/microsoft-ui-xaml/files/4387979/gridsplittertest.zip)

>
| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| November 2019 Update (18363)  | Yes   |

| Device form factor | Saw the problem? |
| :-------------------- | :------------------- |
| Desktop                 | Yes   |

### PR DESCRIPTION
**Pull request by chingucoding**
## Description

This updates the handling of DataTemplateSelectors as discussed in #2170:

1. Call SelectTemplate(IInspectable item)
2. If we received a non nullptr, use that, otherwise try SelectTemplate(IInspectable item, DependencyObject container)
3. Check if that got us a valid pointer, if so use that
4. If we still got a nullptr, throw some exception like "No valid template returned"

Edit: @jtorjo FYI

## Motivation and Context

Fixes #2170
## How Has This Been Tested?

Add 2 new API tests for the cases

### CODE REVIEW

File: dev/Repeater/APITests/ItemTemplateTests.cs [hunk 1]
```diff
@@ -597,6 +598,85 @@ public void ValidateNullItemTemplateAndContainerInItems()
...
+
+                Content.UpdateLayout();
+
+                Verify.AreEqual(numItems, VisualTreeHelper.GetChildrenCount(repeater));
+                for (int i = 0; i < numItems; i++)
+                {
+                    var element = (TextBlock)repeater.TryGetElement(i);
+                    if (i % 2 == 0)
+                    {
+                        // Text is bound to the data for even indicies
```
Comment (ItzLevvie) [inline dev/Repeater/APITests/ItemTemplateTests.cs]:
```suggestion
                        // Text is bound to the data for even indices
```

File: dev/Repeater/APITests/ItemTemplateTests.cs [hunk 2]
```diff
@@ -597,6 +598,85 @@ public void ValidateNullItemTemplateAndContainerInItems()
...
+                        Verify.AreEqual(i.ToString(), element.Text);
+                        Verify.AreEqual(i, element.DataContext);
+                    }
+                    else
+                    {
+                        // Text explicitly set on the element only for odd indicies
```
Comment (ItzLevvie) [inline dev/Repeater/APITests/ItemTemplateTests.cs]:
```suggestion
                        // Text explicitly set on the element only for odd indices
```

**Comment by chingucoding** [comment]
I think the test failures are a problem already tracked by #2160 . Maybe try rerun pipeline?

**Comment by kmahone** [comment]
@chingucoding, we were hitting some issues with test execution. Please merge the latest from master to pick up the fixes. This should resolve the test failures that were being hit.

**Comment by chingucoding** [comment]
This branch is up to date with master.

**Pull request merged**